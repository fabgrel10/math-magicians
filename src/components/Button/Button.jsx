import PropTypes from 'prop-types';

import './Button.module.css';

const Button = ({ value, onClick }) => (
  <button type="button" onClick={onClick}>
    {value}
  </button>
);

Button.propTypes = {
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Button;
