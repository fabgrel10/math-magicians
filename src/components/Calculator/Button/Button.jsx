import React from 'react'

import PropTypes from 'prop-types';

import './Button.module.css';

const Button = ({ onClick, value }) => (
  <button type="button" title="button-title"  onClick={onClick}>
    {value}
  </button>
);

export default Button;
