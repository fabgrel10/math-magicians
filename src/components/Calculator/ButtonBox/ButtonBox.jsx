import React from 'react'

import PropTypes from 'prop-types';

import styles from './ButtonBox.module.css';

const ButtonBox = ({ children }) => <div className={styles.buttonBox} title="button-box-title">{children}</div>;

export default ButtonBox;
