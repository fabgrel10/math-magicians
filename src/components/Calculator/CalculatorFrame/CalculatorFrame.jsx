import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import styles from './CalculatorFrame.module.css';

class CalculatorFrame extends PureComponent {
  render() {
    const { children } = this.props;
    return <div className={styles.wrapper} title="calculator-frame-title">{children}</div>;
  }
}

export default CalculatorFrame;
