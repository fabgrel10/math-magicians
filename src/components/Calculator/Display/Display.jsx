import React from 'react'
import PropTypes from 'prop-types';

import styles from './Display.module.css';

const Display = ({ total, operation, next }) => (
  <div className={styles.display} title="display-title" mode="single" max={70}>
    <p>{next || operation || total || 0}</p>
  </div>
);

Display.propTypes = {
  total: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  operation: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
  ]),
  next: PropTypes.string,
};

Display.defaultProps = {
  total: 0,
  operation: null,
  next: null,
};

export default Display;
