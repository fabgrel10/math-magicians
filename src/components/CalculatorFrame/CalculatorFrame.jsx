import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import styles from './CalculatorFrame.module.css';

class CalculatorFrame extends PureComponent {
  render() {
    const { children } = this.props;
    return <div className={styles.wrapper}>{children}</div>;
  }
}

CalculatorFrame.propTypes = {
  children: PropTypes.node.isRequired,
};

export default CalculatorFrame;
