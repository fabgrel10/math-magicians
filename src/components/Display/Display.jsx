import PropTypes from 'prop-types';

import styles from './Display.module.css';

const Display = ({ result }) => (
  <div className={styles.display} mode="single" max={70}>
    <p>{result}</p>
  </div>
);

Display.propTypes = {
  result: PropTypes.string,
};

Display.defaultProps = {
  result: '0',
};

export default Display;
