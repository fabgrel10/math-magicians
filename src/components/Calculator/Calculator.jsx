import { useState } from 'react';
import CalculatorFrame from '../CalculatorFrame/CalculatorFrame';
import Display from '../Display/Display';
import ButtonBox from '../ButtonBox/ButtonBox';
import Button from '../Button/Button';
import calculate from '../../logic/calculate';

const btnValues = [
  ['AC', '+/-', '%', 'รท'],
  [7, 8, 9, 'x'],
  [4, 5, 6, '-'],
  [1, 2, 3, '+'],
  [0, '.', '='],
];

function Calculator() {
  const [state, setState] = useState({
    total: 0,
    next: '',
    operation: null,
  });

  const handleClick = (btn) => {
    const calc = calculate(state, btn.toString());
    setState(calc);
  };

  const { total, operation, next } = state;

  return (
    <CalculatorFrame>
      <Display total={total} next={next} operation={operation} />
      <ButtonBox>
        {
          btnValues.flat().map((btn, i) => (
            <Button
              // eslint-disable-next-line react/no-array-index-key
              key={i}
              value={btn}
              onClick={() => handleClick(btn)}
            />
          ))
        }
      </ButtonBox>
    </CalculatorFrame>
  );
}

export default Calculator;
