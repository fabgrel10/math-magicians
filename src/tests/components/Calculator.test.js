import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Display from '../../components/Calculator/Display/Display';
import Button from '../../components/Calculator/Button/Button';
import ButtonBox from '../../components/Calculator/ButtonBox/ButtonBox';
import CalculatorFrame from '../../components/Calculator/CalculatorFrame/CalculatorFrame';
import Calculator from '../../components/Calculator/Calculator';

describe('components', () => {
  it('test Display components renders correctly', () => {
    const { queryByTitle } = render(<Display />);
    expect(queryByTitle('display-title')).toBeTruthy();
  });
  it('test Button components renders correctly', () => {
    const { queryByTitle } = render(<Button />);
    expect(queryByTitle('button-title')).toBeTruthy();
  });
  it('test ButtonBox components renders correctly', () => {
    const { queryByTitle } = render(<ButtonBox />);
    expect(queryByTitle('button-box-title')).toBeTruthy();
  });
  it('test CalculatorFrame components renders correctly', () => {
    const { queryByTitle } = render(<CalculatorFrame />);
    expect(queryByTitle('calculator-frame-title')).toBeTruthy();
  });

  it('Calculator component is working', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
