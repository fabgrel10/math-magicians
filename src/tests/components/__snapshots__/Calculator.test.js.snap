// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Calculator component is working 1`] = `
<div
  className="secondary-container"
>
  <div
    className="wrapper"
  >
    <div
      className="column"
    >
       
      <h1
        className="title"
      >
        Let’s do some math!
      </h1>
    </div>
    <div
      className="column"
    >
      <div
        className="wrapper"
        title="calculator-frame-title"
      >
        <div
          className="display"
          max={70}
          mode="single"
          title="display-title"
        >
          <p>
            0
          </p>
        </div>
        <div
          className="buttonBox"
          title="button-box-title"
        >
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            AC
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            +/-
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            %
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            ÷
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            7
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            8
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            9
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            x
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            4
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            5
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            6
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            -
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            1
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            2
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            3
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            +
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            0
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            .
          </button>
          <button
            onClick={[Function]}
            title="button-title"
            type="button"
          >
            =
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
